### PR TITLE
[coroutines] Introduce [[clang::coro_lifetimebound]]

### PR DESCRIPTION
**Pull request by usx95**
Adds attribute `[[clang::coro_lifetimebound]]`.

All arguments to a function are considered to be **lifetime bound** if the function 
returns a type annotated with ``[[clang::coro_lifetimebound]]`` and ``[[clang::coro_return_type]]``.

### CODE REVIEW

File: clang/include/clang/Basic/AttrDocs.td [hunk 1]
```diff
@@ -7483,7 +7483,6 @@ generation of the other destruction cases, optimizing the above `foo.destroy` to
   }];
 }
 
-
```
Comment (ChuanqiXu9) [inline clang/include/clang/Basic/AttrDocs.td:7486]: nit: unnecessary changes.

File: clang/lib/Sema/SemaInit.cpp [hunk 2]
```diff
@@ -7580,10 +7580,15 @@ static void visitLifetimeBoundArguments(IndirectLocalPath &Path, Expr *Call,
   if (ObjectArg && implicitObjectParamIsLifetimeBound(Callee))
     VisitLifetimeBoundArg(Callee, ObjectArg);
 
+  bool checkCoroCall = false;
+  if (const auto *RD = Callee->getReturnType()->getAsRecordDecl()) {
+    checkCoroCall |= RD->hasAttr<CoroLifetimeBoundAttr>() &&
```
Comment (ChuanqiXu9) [inline clang/lib/Sema/SemaInit.cpp:7585]:
```suggestion
    checkCoroCall = RD->hasAttr<CoroLifetimeBoundAttr>() &&
```

nit:

**Review by ChuanqiXu9 — APPROVED** [review]
LGTM. Please wait for few days in case there are other comments.

File: clang/include/clang/Basic/AttrDocs.td (hunk 1 above)
Comment (usx95) [reply clang/include/clang/Basic/AttrDocs.td:7486]: Acknowledged. Keeping it to match the 1 line space style of the file.

File: clang/lib/Sema/SemaInit.cpp [hunk 3]
```diff
@@ -7580,10 +7580,15 @@ static void visitLifetimeBoundArguments(IndirectLocalPath &Path, Expr *Call,
   if (ObjectArg && implicitObjectParamIsLifetimeBound(Callee))
     VisitLifetimeBoundArg(Callee, ObjectArg);
 
+  bool checkCoroCall = false;
```
Comment (hokein) [inline clang/lib/Sema/SemaInit.cpp:7583]: nit: `CheckCoroCall`, LLVM uses `CamelCase` naming style for local variables.

File: clang/include/clang/Basic/AttrDocs.td [hunk 4]
```diff
@@ -7581,3 +7580,59 @@ alignment boundary. Its value must be a power of 2, between 1 and 4096
...
+  }
+  Task<int> temporary_reference() {
+    auto foo = coro(1); // warning: capturing reference to a temporary which would die after the expression.
+
+    int a = 1;
+    auto bar = coro_wrapper(a, 0); // warning: `b` captures reference to a temporary.
+
+    co_return co_await coro(1); // fine.
+  }
+  Task<int> stack_reference(int a) {
```
Comment (hokein) [inline clang/include/clang/Basic/AttrDocs.td:7634]: nit: looks like `[[clang::coro_wrapper]]` attr is missing.

File: clang/test/SemaCXX/coro-lifetimebound.cpp [hunk 5]
```diff
@@ -0,0 +1,120 @@
...
+    }
+    suspend_always initial_suspend();
+    suspend_always final_suspend() noexcept;
+    void unhandled_exception();
+    void return_value(const T &t);
+  };
+};
+
+Co<int> foo_coro(const int& a) { co_return a; }
+Co<int> bar(int a) { return foo_coro(a); }
```
Comment (hokein) [inline clang/test/SemaCXX/coro-lifetimebound.cpp:119]: nit: I think we'd better add the `coro_wrapper` attr to this function. If the `Co` was a CRT, the clang will emit a hard error, and this error is irrelevant.

File: clang/test/SemaCXX/coro-lifetimebound.cpp [hunk 6]
```diff
@@ -0,0 +1,120 @@
...
+Co<int> value_coro(int b) { co_return co_await foo_coro(b); }
+[[clang::coro_wrapper]] Co<int> wrapper1(int b) { return value_coro(b); }
+[[clang::coro_wrapper]] Co<int> wrapper2(const int& b) { return value_coro(b); }
+}
+
+// =============================================================================
+// Lifetime bound but not a Coroutine Return Type: No analysis.
+// =============================================================================
+namespace not_a_crt {
+template <typename T> struct [[clang::coro_lifetimebound]] Co {
```
Comment (hokein) [inline clang/test/SemaCXX/coro-lifetimebound.cpp:106]: nit: consider changing the name to `CoNoCRT` (we have used the same name in the above CRT tests).

**Review by hokein — APPROVED** [review]
Thanks, looks good from my side. Some nits.

File: clang/include/clang/Basic/AttrDocs.td (hunk 4 above)
Comment (usx95) [reply clang/include/clang/Basic/AttrDocs.td:7634]: Thanks. Good catch.

File: clang/test/SemaCXX/coro-lifetimebound.cpp (hunk 5 above)
Comment (usx95) [reply clang/test/SemaCXX/coro-lifetimebound.cpp:119]: Made this a coroutine instead. I do not think `coro_wrapper` is meaningful without `coro_return_type`.

**Review by ilya-biryukov — APPROVED** [review]
LGTM with a few suggestiosn.

One is in the comment.
I also suggest to add a warning when users mark the type as `coro_lifetimebound` without `coro_return_type`.
The analysis will be disabled in this case and it may be hard to understand why.

However, I don't think that's a reason to block this change. This could be a follow-up.

**Comment by usx95** [comment]
I would add the mentioned warning in a follow-up change.
